rclcpp/doc/notes_on_statically_typed_parameters.md

```diff
@@ -0,0 +1,111 @@
+# Notes on statically typed parameters
+
+## Introduction
+
+Up to ROS 2 Foxy, all parameters could be changed of type anytime, except the user installed a parameter callback to reject that change.
+This could generate confusing errors, for example:
+
+```
+$ ros2 run demo_nodes_py listener &
+$ ros2 param set /listener use_sim_time not_a_boolean
+[ERROR] [1614712713.233733147] [listener]: use_sim_time parameter set to something besides a bool
+Set parameter successful
+$ ros2 param get /listener use_sim_time
+String value is: not_a_boolean
+```
+
+For most use cases, having a static parameter types is enough.
+This notes document some of the decisions that were made when implementing static parameter types enforcement in:
+
+* https://github.com/ros2/rclcpp/pull/1522
+* https://github.com/ros2/rclpy/pull/683
+
+## Allowing dynamically typed parameters
+
+There might be some scenarios were dynamic typing is desired, so a `dynamic_typing` field was added to the [parameter descriptor](https://github.com/ros2/rcl_interfaces/blob/09b5ed93a733adb9deddc47f9a4a8c6e9f584667/rcl_interfaces/msg/ParameterDescriptor.msg#L25).
+It defaults to `false`.
+
+For example:
+
+```cpp
+rcl_interfaces::msg::ParameterDescriptor descriptor;
+descriptor.dynamic_typing = true;
+
+node->declare_parameter("dynamically_typed_parameter", rclcpp::ParameterValue{}, descriptor);
+```
+
+```py
+rcl_interfaces.msg.ParameterDescriptor descriptor;
+descriptor.dynamic_typing = True;
+
+node.declare_parameter("dynamically_typed_parameter", None, descriptor);
+```
+
+## How is the parameter type specified?
+
+The parameter type will be inferred from the default value provided when declaring it.
+
+## Statically typed parameters when allowing undeclared parameters
+
+When undeclared parameters are allowed, the descriptor of undeclared parameters will have the dynamic typing field set.
+That's because in that case there is no declaration, and it doesn't make much sense to enforce the type of the first value set.
```

Comment:
i think this makes sense to me.

Reply:
yeah, my question is that if you actually made a declaration e.g. `node->declare_parameter("my_int", 5)`, should an static type being enforced even if `allow_undeclared_parameters == true`?

Reply:
I could be wrong, but when `allow_undeclared_parameters` is true I think `get_parameter` for something that was never set or declared will return `NOT_SET`. If that's the case, that would be a reason to justify this beyond "it doesn't make much sense" which is pretty vague.

Reply:
> I could be wrong, but when allow_undeclared_parameters is true I think get_parameter for something that was never set or declared will return NOT_SET. If that's the case, that would be a reason to justify this beyond "it doesn't make much sense" which is pretty vague.

That sounds fine to me

Reply:
Fixed in https://github.com/ros2/rclcpp/pull/1568/commits/f245d42a871e4f33bc075cfd9f7d8994c05476c8.